### ISSUE TITLE
Add algorithm for DeepWalk 

### ISSUE THREAD

**Issue by czgdp1807** [codezonediitj/pydatastructs#65]
#### Description of the problem

Once we are done adding basic framework for Graphs, we can start thinking of adding DeepWalk algorithm. 
We first need to decide the feasibility of implementation of the above algorithm. 
#### References/Other comments
[1] https://github.com/phanein/deepwalk

**Comment by Saptashrungi** [comment]
Is this issue up for implementation? What are basic requirements in case of graphs?

**Comment by czgdp1807** [comment]
One has to read this paper, https://arxiv.org/pdf/1403.6652.pdf and see how the current Graph API in `master` can be used or what modifications are needed to implement this algorithm. This issue will require a series of PRs with different components added in each to get the algorithm implemented successfully.

**Comment by Saptashrungi** [comment]
Okay, I will give it try.

**Comment by Saptashrungi** [comment]
I would like to work on this issue as it seems interesting to me. I read the paper as well as went through the algorithm. I understood some of it and at some places I am stuck. For eg:

Φ∈R|V|×d:  Initialization:  Sample Φ from U|V|×d

What is U here? 
d is embedding size.
The sample Φ is a vector space I assume.

I will soon make a PR with a rough code. It will also contain code for random_walk. This function does not however implement Random Walk with Restart (RWR), as the paper you provided said that restarts are not necessary for deep walk. I would like to discuss whether or not to include restarts in random walk function, if necessary we can modify the code accordingly.
I haven't searched about SkipGram. I will do it next.

**Comment by czgdp1807** [comment]
> I will soon make a PR with a rough code.

PR will be too early, without having discussions on design. We don't even know whether the APIs for graphs will be enough for implementing this. We need to pick the simplest algorithm in the paper(if there are multiple) and discuss on it first.

**Comment by czgdp1807** [comment]
In fact, they have used the following as inputs to the algorithm,

1. *w*, window size 
2. *d*, embedding size
3. *g*, walks per vertex
4. *t*, walk length

What all do they mean and how can we represent them in pydatastructs?

**Comment by Saptashrungi** [comment]
I am still reading about them. For now only Random Walk can be implemented
using AdjacencyList.

**Comment by czgdp1807** [comment]
> I am still reading about them

I think first that part should be clarified as they are going to be at the entry points of the deep walk algorithm.

**Comment by Saptashrungi** [comment]
Okay.

**Comment by Saptashrungi** [comment]
Window size is the basically the size of the surrounding area of the node which we are considering for node embedding in a graph. We can set the window size according to our convenience.
For example if its value is x, then we can make an edge list where we represent the node and the neighbors associated with node under consideration in the given window size.

**Comment by czgdp1807** [comment]
It looks like all the variables in https://github.com/codezonediitj/pydatastructs/issues/65#issuecomment-602203419 are integers, so no extra framework is needed to support these values?

**Comment by Saptashrungi** [comment]
Yes. The algorithm can be implemented with existing framework.

**Comment by Saptashrungi** [comment]
For Skipgram implementation Word2Vec model is generally used. It is google’s pre-trained model. It contains word vectors for a vocabulary of 3 million words trained on around 100 billion words from the google news dataset. It is a 1.5 GB download. We need an alternative for this model.

**Comment by czgdp1807** [comment]
> We first need to decide the feasibility of implementation of the above algorithm.

The problems which we will face,

1. `DeepWalk` relates to graphs but not in a way which should keep the algorithm in the scope of this project.
2. `SkipGram` is an optimisation algorithm which is again doesn't suit the repository.
3. `RandomWalk` can be implemented only if there are enough use cases for the following outpu i.e., the transcript of a random walk.
> While the transcript (the list of vertices in the order they are
visited) of a particular random walk is sometimes of interest

The issue title can be narrowed down to `RandomWalk` but still the question of feasibility remains to be answered before moving towards the first step i.e., API design.

**Comment by Saptashrungi** [comment]
According to me the RandomWalk algorithm can be implemented with existing framework.
I read the source provided by you for the algorithm, will it be enough to implement till section 10.3, which is feasible.

**Comment by czgdp1807** [comment]
> 3\. `RandomWalk` can be implemented only if there are enough use cases for the following output i.e., the transcript of a random walk.

First this should be done.

**Comment by Saptashrungi** [comment]
`RandomWalk` is used in useful algorithms like `node2vec` and `graph2vec` algorithms, that create node embeddings. It can be used as part of the `Walktrap` and `Infomap` community detection algorithms. If a random walk returns a small set of nodes repeatedly, then it indicates that those set of nodes may have a community structure. 
We can implement node2vec using `RandomWalk` later.
[https://arxiv.org/pdf/1607.00653.pdf](https://arxiv.org/pdf/1607.00653.pdf)

**Comment by czgdp1807** [comment]
> It can be used as part of the node2vec and graph2vec algorithms, that create node embeddings.
It can be used as part of the Walktrap and Infomap community detection algorithms. If a random walk returns a small set of nodes repeatedly, then it indicates that those set of nodes may have a community structure.

Quoting from https://neo4j.com/docs/graph-algorithms/current/labs-algorithms/random-walk/

The above is not what is expected as a "use case" for implementing random walks in this project. 

I came across [Wilson's algorithm](https://bl.ocks.org/mbostock/11357811) which uses [loop-erased random walks](https://en.wikipedia.org/wiki/Loop-erased_random_walk) as a sub-algorithm. It generates uniform spanning tree. This can be a good reason to get random walks and loop erased random walks implemented as this is where we can use their trace/transcript.

**Comment by Saptashrungi** [comment]
So we need to implement loop-erased randomwalks first. I am working on it.

**Comment by czgdp1807** [comment]
Both are required. First simple random walks and then loop erased random walks.

**Comment by JyotiSunkara** [comment]
Hello, I have implemented a basic random walk code in python. Done for the drunk man's random walk mathematical problem. Can be seen here: https://github.com/JyotiSunkara/Drunk-Man-Random-Walk
Please assign this issue to me, so that I can work on it.

**Comment by czgdp1807** [comment]
@JyotiSunkara Thanks for sharing your code. According to our [issue policy]() we don't assign issues. As a first step please design an API for simple random walks. The input can be a graph and a starting node and may be some indicator input(probably an the length of output required) to let the function know when to stop. The output can be a list of nodes traversed by the random walk. The code should go to, `pydatastructs.graphs.algorithms.py`.

**Comment by RidhimaKohli** [comment]
Hello , I am Ridhima , working with Saptashrungi . I have shared some code regarding this issue

**Comment by JyotiSunkara** [comment]
For the API for random walks, I think the input  should be 
`n: The number of steps taken in the random walk`
`d: The dimensions of the random walk. Can be 1/2/3D`
`o: The start point or the origin of the random walk`

The out put can be a list of coordinates/nodes visited and the corresponding graph.
Is this okay?

**Comment by czgdp1807** [comment]
> d: The dimensions of the random walk. Can be 1/2/3D

How will you explain the concept of 2D random walks on Graphs?

**Comment by JyotiSunkara** [comment]
In the graph each point (x, y) will be a position visited during the walk. (Time will not be represented)

**Comment by czgdp1807** [comment]
Well, IMHO, 1D/2D/3D random walks in the context of Graph data structure don't hold any meaning. [This document](https://people.math.osu.edu/husen.1/teaching/571/random_walks.pdf) reflects what is required to be implemented. Though what they have done is similar to Markov chains i.e., they made a transition matrix with M[i][j] denotes the probability that the transition will be made to vertex v_j from v_i. However, we will do that implicitly. Please let me know if you have any questions.

**Comment by JyotiSunkara** [comment]
A particle moving on the surface of a fluid exhibits 2D random walk and its trajectory can be represented by the graph we were talking about. (I was talking about a matplotlib graph)

I have understood the document and the requirement now.
I think the API should be,
Input:
`G(V, E)`: The graph 
`s`: Start node 

`L`: List of nodes traversed
`P`: List of probabilitiy of reaching each node

**Comment by Saptashrungi** [comment]
[This document](https://people.math.osu.edu/husen.1/teaching/571/random_walks.pdf) 
The probability of next node in random walk can be decided based on the weight of the edge or degree according to the source you shared. Which one should be implemented?

**Comment by czgdp1807** [comment]
ping @JyotiSunkara 

> A particle moving on the surface of a fluid exhibits 2D random walk and its trajectory can be represented by the graph we were talking about. (I was talking about a matplotlib graph)

That's not in the scope of this package, IMHO.



> `G(V, E)`: The graph

graph (See `Graph` class in `graph.py`)

> `s`: Start node

s -> start_node (Use full names, tab completion is there so lengthy names not a problem)

Add an extra parameter to signal the function when to stop. May be the maximum required length of the output list. 

Before starting work just go through the code in [this file](https://github.com/codezonediitj/pydatastructs/blob/master/pydatastructs/graphs/algorithms.py).

**Comment by czgdp1807** [comment]
> The probability of next node in random walk can be decided based on the weight of the edge or degree according to the source you shared. Which one should be implemented?

Well, providing an option to the user to decide will be better. By default, next vertex can be chosen uniformly from the set of the neighbours of current vertex.

**Comment by Saptashrungi** [comment]
> Well, providing an option to the user to decide will be better. By default, next vertex can be chosen uniformly from the set of the neighbours of current vertex.

If we keep options then user will be provided the following options:
1. Uniform `Randomwalk` 
2. `Randomwalk` with probability based on weight of edges 
3. `Randomwalk` with probability based on degree of current node

The user will give the following inputs:
1. Graph : `AdjacencyList` or `AdjacencyMatrix` (option will be provided)
2. `Type` of Randomwalk which will decide the `probability` for next node
3. `path_length`
4. `no_of_walks` 
5. `start` (optional)

The output will be list of nodes traversed by the `random_walk` function.
Any suggestions for the above API?

Sources:
1. Last section of: [RandomWalk with Weighted Graphs](http://www.lirmm.fr/~sau/JCALM/Josep.pdf)

**Comment by czgdp1807** [comment]
> If we keep options then user will be provided the following options:
> 
> 1. Uniform `Randomwalk`
> 2. `Randomwalk` with probability based on weight of edges
> 3. `Randomwalk` with probability based on degree of current node

How will those options provided through code?




> 1. Graph : `AdjacencyList` or `AdjacencyMatrix` (option will be provided)

Not needed. Can be figured out from the `graph` which is passed in the arguments itself.



> 2\. `Type` of Randomwalk which will decide the `probability` for next node

How? What will be the values that can be attained by `Type` and what will be the type of `Type`?





> 5\. `start` (optional)

If start is not given then from where will you start the walk?

**Comment by Saptashrungi** [comment]
If user does not provide the start node, the function will randomly start a walk from a randomly chosen node.
`Type` will be string, where it can be : "Uniform", "Weight_Biased", "Random"

The code for all the types would be similar. Only difference will be in the `probability` list.
Based on the type provided by the user will can accordingly create `probability` of data-type list.

**Comment by czgdp1807** [comment]
If observed there is no difference between uniformly random and degree based probability. So, if there should be only two options either weight based probabilities or degree based probabilities. By default it should be degree based.

**Comment by Saptashrungi** [comment]
> If observed there is no difference between uniformly random and degree based probability. So, if there should be only two options either weight based probabilities or degree based probabilities. By default it should be degree based.

Okay, we will give two options to the user, keeping degree based as default. Rest of the API would be the same. Should I start working on it?

**Comment by czgdp1807** [comment]
Where are the weight based probabilities used? Degree based are used in Wilson's algorithm but what about the other ones? If we don't have any reason to include weight based probabilities then no need to complicate the things.

**Comment by Saptashrungi** [comment]
I tried but was unable to find a suitable algorithm for random walk with weights. I will implement degree based random walk and make a PR soon. Any more suggestions? 
How do we write tests for this, as the output will be random?

**Comment by czgdp1807** [comment]
Fix the seed in the test function to get deterministic outputs.

**Comment by JyotiSunkara** [comment]
> ping @JyotiSunkara
> 
> > A particle moving on the surface of a fluid exhibits 2D random walk and its trajectory can be represented by the graph we were talking about. (I was talking about a matplotlib graph)
> 
> That's not in the scope of this package, IMHO.
> 
> > `G(V, E)`: The graph
> 
> graph (See `Graph` class in `graph.py`)
> 
> > `s`: Start node
> 
> s -> start_node (Use full names, tab completion is there so lengthy names not a problem)
> 
> Add an extra parameter to signal the function when to stop. May be the maximum required length of the output list.
> 
> > P: List of probabilitiy of reaching each node
> 
> Sounds creative.
> 
> Before starting work just go through the code in [this file](https://github.com/codezonediitj/pydatastructs/blob/master/pydatastructs/graphs/algorithms.py).

Understood, will code and ask for feedback in while.

**Comment by mridul45** [comment]
I would be happy to contribute to the project as I find my skills a good fit for the issue . As a GSSOC 23 member I humbly ask to allow me to contribute to the issue. If the issue is still available
